File: src/waypoint/platform/mock.py
"""In-memory mock backend used by tests and by GUI/CLI development without
real hardware. This is what keeps core/engine logic testable on Linux CI
even though the primary target platform is Windows.
"""

from __future__ import annotations

from waypoint.core.models import Device
from waypoint.platform.base import InstallResult


class MockDeviceBackend:
    def __init__(self, devices: list[Device] | None = None) -> None:
        self._devices = devices or []
        self.backup_calls: list[tuple[str, str]] = []
        self.install_calls: list[tuple[str, str, bool]] = []
        self.restore_point_calls: list[str] = []
        self.restore_point_should_succeed = True

    def enumerate_devices(self) -> list[Device]:
        return list(self._devices)

    def export_driver_backup(self, device: Device, dest_dir: str) -> str:
        self.backup_calls.append((device.instance_id, dest_dir))
        return f"{dest_dir}/{device.instance_id}.backup"

    def install_driver(self, device: Device, package_path: str, *, dry_run: bool) -> InstallResult:
        self.install_calls.append((device.instance_id, package_path, dry_run))
        command = f"pnputil /add-driver {package_path} /install"
        if dry_run:
            return InstallResult(success=True, commands_run=[command], message="dry-run: no changes made")
        return InstallResult(success=True, commands_run=[command], message="installed")

    def create_restore_point(self, description: str) -> bool:
        self.restore_point_calls.append(description)
        return self.restore_point_should_succeed

File: rust/crates/platform/src/mock.rs
//! In-memory mock backend used by tests and by GUI/CLI development without
//! real hardware. Direct port of `platform/mock.py`. This is what keeps
//! core/engine logic testable on Linux CI even though the primary target
//! platform is Windows.

use std::cell::RefCell;

use waypoint_core::Device;

use crate::base::{DeviceBackend, InstallResult};

pub struct MockDeviceBackend {
    devices: Vec<Device>,
    pub backup_calls: RefCell<Vec<(String, String)>>,
    pub install_calls: RefCell<Vec<(String, String, bool)>>,
    pub restore_point_calls: RefCell<Vec<String>>,
    pub restore_point_should_succeed: bool,
}

impl MockDeviceBackend {
    pub fn new(devices: Vec<Device>) -> Self {
        Self {
            devices,
            backup_calls: RefCell::new(Vec::new()),
            install_calls: RefCell::new(Vec::new()),
            restore_point_calls: RefCell::new(Vec::new()),
            restore_point_should_succeed: true,
        }
    }
}

impl DeviceBackend for MockDeviceBackend {
    fn name(&self) -> &'static str {
        "MockDeviceBackend"
    }

    fn enumerate_devices(&self) -> Result<Vec<Device>, String> {
        Ok(self.devices.clone())
    }

    fn export_driver_backup(&self, device: &Device, dest_dir: &str) -> Result<String, String> {
        self.backup_calls
            .borrow_mut()
            .push((device.instance_id.clone(), dest_dir.to_string()));
        Ok(format!("{dest_dir}/{}.backup", device.instance_id))
    }

    fn install_driver(&self, device: &Device, package_path: &str, dry_run: bool) -> InstallResult {
        self.install_calls.borrow_mut().push((
            device.instance_id.clone(),
            package_path.to_string(),
            dry_run,
        ));
        let command = format!("pnputil /add-driver {package_path} /install");
        if dry_run {
            InstallResult {
                success: true,
                commands_run: vec![command],
                message: "dry-run: no changes made".to_string(),
            }
        } else {
            InstallResult {
                success: true,
                commands_run: vec![command],
                message: "installed".to_string(),
            }
        }
    }

    fn create_restore_point(&self, description: &str) -> bool {
        self.restore_point_calls
            .borrow_mut()
            .push(description.to_string());
        self.restore_point_should_succeed
    }
}
